### PR TITLE
Server-Side Logging for Unofficial Client Connections

### PR DESCRIPTION
**Pull request by rafia-k**
*Adding logging for when a user connects to a Kafka server (on ApiVersionsRequest) with a client other than one of the LinkedIn official clients.*

*No tests added.*

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/LogConfig.scala [hunk 1]
```diff
@@ -66,6 +66,7 @@ object Defaults {
   val MaxIdMapSnapshots = kafka.server.Defaults.MaxIdMapSnapshots
   val MessageDownConversionEnable = kafka.server.Defaults.MessageDownConversionEnable
   val ProducerBatchDecompressionEnable = kafka.server.Defaults.ProducerBatchDecompressionEnable
+  val UnofficialClientLoggingEnable = kafka.server.Defaults.UnofficialClientLoggingEnable
```
Comment (kehuum) [inline core/src/main/scala/kafka/log/LogConfig.scala]: this config has nothing to do with Log, why is it here? Log here means the log segments stored in kafka broker, not the log4j logging config.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -1683,8 +1686,48 @@ class KafkaApis(val requestChannel: RequestChannel,
     // If this is considered to leak information about the broker version a workaround is to use SSL
     // with client authentication which is performed at an earlier stage of the connection where the
     // ApiVersionRequest is not available.
+    val apiVersionRequest = request.body[ApiVersionsRequest]
+
+    // flush unofficial client set every hour
+    if (System.currentTimeMillis() - flushStartTime > unofficialClientSetTtl) {
+      unofficialClientIdentities.clear()
```
Comment (kehuum) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: we should not clear the entire cache every hour, instead, use a guava cache with TTL or similar to evict entries that reached TTL.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 3]
```diff
@@ -1683,8 +1686,48 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+      flushStartTime = System.currentTimeMillis()
+    }
+
+    if (config.unofficialClientLoggingEnable) {
+      val softwareName = apiVersionRequest.data.clientSoftwareName()
+      val clientIdentity = request.context.clientId() + "-" + request.context.clientAddress() + "-" + request.context.principal()
+      val UNOFFICIAL_CLIENT = "UNOFFICIAL"
+
+      def getClientTypeFromClientSoftwareName(softwareName: String): String = {
```
Comment (kehuum) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: we only intend to log unofficial clients, so why do we care about which type of clients like avro/raw/admin? is it possible if we just check if (softwareName == defaultName) -> unofficial clients?

Comment (rafia-k) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: It seems that there are four default names: "apache-kafka-java", "linkedin-kafka-java", "li-oss-consumer-java", and "li-oss-producer-java". Based on that would it still be better to check for each of these than the current way?

Comment (rafia-k) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Implemented with the four default names for now

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 4]
```diff
@@ -111,6 +111,15 @@ class KafkaApis(val requestChannel: RequestChannel,
   val adminZkClient = new AdminZkClient(zkClient)
   private val alterAclsPurgatory = new DelayedFuturePurgatory(purgatoryName = "AlterAcls", brokerId = config.brokerId)
 
+  val unofficialClientsCache: LoadingCache[String, String] = CacheBuilder.newBuilder()
+    .expireAfterWrite(1, TimeUnit.HOURS)
```
Comment (kehuum) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: we might also want to make time to expire a config

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 5]
```diff
@@ -72,6 +72,7 @@ object Defaults {
   val ProducerBatchDecompressionEnable = true
   val PreferredController = false
   val AllowPreferredControllerFallback = true
+  val UnofficialClientLoggingEnable = true
```
Comment (kehuum) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: normally for newly added features the default value should be false, i.e. by default the feature is turned off.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 6]
```diff
@@ -675,6 +678,8 @@ object KafkaConfig {
   val AllowPreferredControllerFallbackDoc = "Specifies whether a non-preferred controller node (broker) is allowed to become the controller." +
   " This configuration is expected to be configured at cluster level via dynamic broker configuration to provide a consistent configuration among all brokers." +
   " If AllowPreferredControllerFallback is dynamically set to false and there is no preferred controllers, the non-preferred active controller does not resign."
+  val UnofficialClientLoggingEnableDoc = "Controls whether logging occurs when an ApiVersionsRequest is received from an unofficial client."
```
Comment (kehuum) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: add more description on what is unofficial client? i.e. clients instantiated from unsupported ways at LI like directly from open source clients etc.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 4 above)
Comment (rafia-k) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: would making it more granular, such as minutes, be a good idea?

Comment (kehuum) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: hours is sufficient imo

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 7]
```diff
@@ -111,6 +111,15 @@ class KafkaApis(val requestChannel: RequestChannel,
   val adminZkClient = new AdminZkClient(zkClient)
   private val alterAclsPurgatory = new DelayedFuturePurgatory(purgatoryName = "AlterAcls", brokerId = config.brokerId)
 
+  val unofficialClientsCache: LoadingCache[String, String] = CacheBuilder.newBuilder()
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/server/KafkaApis.scala:114]: nit - youre not really fetching any remote initial state per client, just initializing it to empty. does this need to be a loading cache?

Comment (rafia-k) [reply core/src/main/scala/kafka/server/KafkaApis.scala:114]: I'm not sure that I understand. the documentation recommends using a LoadingCache as opposed to a Cache if there's a sensible default function to load/compute a value for a key (https://github.com/google/guava/wiki/CachesExplained#population), which in this case I believe would be the clientId itself. is this what you're referring to?

Comment (radai-rosenblatt) [reply core/src/main/scala/kafka/server/KafkaApis.scala:114]: ok, fine with me

**Review by radai-rosenblatt — APPROVED** [review]

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 8]
```diff
@@ -1683,8 +1692,20 @@ class KafkaApis(val requestChannel: RequestChannel,
     // If this is considered to leak information about the broker version a workaround is to use SSL
     // with client authentication which is performed at an earlier stage of the connection where the
     // ApiVersionRequest is not available.
+    val apiVersionRequest = request.body[ApiVersionsRequest]
+
+    if (config.unofficialClientLoggingEnable) {
+      val softwareName = apiVersionRequest.data.clientSoftwareName()
+      val clientIdentity = request.context.clientId() + " " + request.context.clientAddress() + " " + request.context.principal()
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit - this string does not need to be calculated until isUnofficial is true

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 9]
```diff
@@ -1683,8 +1692,20 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+      val isUnofficial = softwareName.equals("linkedin-kafka-java") || softwareName.equals("apache-kafka-java") || softwareName.equals("li-oss-producer-java") || softwareName.equals("li-oss-consumer-java")
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: replace this with a "Set<String> clientsWeLike" and make this a contains() call

Comment (radai-rosenblatt) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: may also be worth adding a config, "clientsWeLike" that accepts a CSV

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 10]
```diff
@@ -72,6 +72,9 @@ object Defaults {
   val ProducerBatchDecompressionEnable = true
   val PreferredController = false
   val AllowPreferredControllerFallback = true
+  val UnofficialClientLoggingEnable = false
+  val UnofficialClientCacheTtl = 1
+  val UnofficialClientSoftwareNames = util.Arrays.asList("linkedin-kafka-java", "apache-kafka-java", "li-oss-producer-java", "li-oss-consumer-java")
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: these are the official ones though ?

**Review by kehuum — APPROVED** [review]

**Pull request merged**